### PR TITLE
Practices update: Javascript and Sql Alchemy

### PR DESCRIPTION
**Pull request by rrees**
This moves some of our advice up from the project-level to general team advice that applies irrespective of which project we are working on.

For the most part this is not new but I don't think we've written down the SQL Alchemy stuff before.

### CODE REVIEW

File: practices/core-technologies/sql-alchemy-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+---
+layout: page
+
+title: SQL Alchemy Guideline
+
+show_toc: true
+
+---
+
+## Query types
```
Comment (jneves) [inline practices/core-technologies/sql-alchemy-guidelines.md]: Disagree. There are different use cases for `one`, `one_or_none` and `first`. If you only want one element, I'd say it's more pythonic to just use `one` and deal with the exceptions appropriately. Most of the uses of `first` are hacks for when the code assumes only one instance exists but the data model doesn't enforce it, so it's actually a bad smell.

File: practices/core-technologies/sql-alchemy-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+See also the advice in the Essential SQL Alchemy book.
+
+## Clauses
+
+Prefer IN (`in_`) over OR
+
+### Rationale
+
+In the vast majority of cases using the IN clause will be quicker and easier for the query planner to optimise.
```
Comment (jneves) [inline practices/core-technologies/sql-alchemy-guidelines.md]: This needs an addition that the list on the `IN` needs to be checked for never being empty, otherwise the query becomes really slow.

File: practices/core-technologies/javascript-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+---
+layout: page
+title: Javascript Guidelines
+
+---
+
+## Prefer null to undefined
```
Comment (jstadnik) [inline practices/core-technologies/javascript-guidelines.md]: As confirmed with a friendly neighbourhood English speaker, Olly, don't you mean "Prefer undefined to null" here?

File: practices/core-technologies/sql-alchemy-guidelines.md (hunk 1 above)
Comment (jstadnik) [reply practices/core-technologies/sql-alchemy-guidelines.md]: I would say I agree with Joao here -- I think blanket switch to "first" could end up hiding problems with the data and therefore take us a while to notice something is wrong.

File: practices/core-technologies/sql-alchemy-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+The reason has to do with the amount of work the query engine has to do to determine the possible combinations whereas the IN will often define a small subset of the data using indexed values.
+
+See also [Avoiding OR for better query performance](https://www.cybertec-postgresql.com/en/avoid-or-for-better-performance/)
```
Comment (jstadnik) [inline practices/core-technologies/sql-alchemy-guidelines.md]: :)

File: practices/core-technologies/sql-alchemy-guidelines.md (hunk 1 above)
Comment (jstadnik) [reply practices/core-technologies/sql-alchemy-guidelines.md]: I would say I agree with Joao here -- I think blanket switch to "first" could end up hiding problems with the data and therefore take us a while to notice something is wrong.

File: practices/core-technologies/javascript-guidelines.md (hunk 3 above)
Comment (rrees) [reply practices/core-technologies/javascript-guidelines.md]: Correct

File: practices/core-technologies/sql-alchemy-guidelines.md (hunk 1 above)
Comment (rrees) [reply practices/core-technologies/sql-alchemy-guidelines.md]: The way we are using `one` at the moment does not address the issues with data structure as we're not handling the exceptions correctly. I can amend this advice to do that but I think it's important to stop kidding ourselves that `one` is somehow telling us about whether our data is "good" or "bad". Our actual implementation is relying on the implicit existence of primary keys on the underlying tables.

Comment (rrees) [reply practices/core-technologies/sql-alchemy-guidelines.md]: @jneves the Pythonic argument doesn't work for me if there is a published recommendation in the technology community to favour `first` and there isn't meaningful exception handling. However I acknowledge that the Python community is split on the use of exceptions.

File: practices/core-technologies/sql-alchemy-guidelines.md (hunk 2 above)
Comment (rrees) [reply practices/core-technologies/sql-alchemy-guidelines.md]: Isn't that equally true of OR?

**Comment by rrees** [comment]
@Amwam can you take a look at this one again, I'd like to try and get it updated and merged so we can start talking about the changes in SQL Alchemy 1.4

File: practices/databases/sql-alchemy-guidelines.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+Use of the ResultProxy iterable creates less memory-pressure than `all` and in most cases we don't want or need to do anything with all the results for the query in one go in the application.
+
+`first` is easier to understand for maintenance as it returns `None` if no row is found.
+
+
+If you use `one` then multiple rows or no rows found are both exceptions and in our code we rarely handle those exceptions explicitly so it is hard to follow what the code should be doing in those cases.
+
+For `one_or_none` then you should do something appropriate with the multiple rows exception.
+
+The `fetch` methods may require slightly different management of connections so they should probably never be used in our codebases.
```
Comment (Amwam) [inline practices/databases/sql-alchemy-guidelines.md:31]: Looking at the docs fetchone/fetchmany are only really available for compatibility with v1, so maybe thats worth calling out a bit more.
https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=fetch#sqlalchemy.engine.Result.fetchone

File: practices/databases/sql-alchemy-guidelines.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+See also the advice in the Essential SQL Alchemy book.
```
Comment (Amwam) [inline practices/databases/sql-alchemy-guidelines.md:33]:
```suggestion
See also the advice in the Essential SQLAlchemy book.
```

**Review by Amwam — APPROVED** [review]

File: practices/databases/sql-alchemy-guidelines.md (hunk 5 above)
Comment (rrees) [reply practices/databases/sql-alchemy-guidelines.md:31]: Let's update this in light of 1.4/2.0 once it's merged are we thinking of switching to prohibition of use (the link is currently blocked by Bitglass 🙄
